finality-aleph/src/justification/compatibility.rs

```diff
@@ -57,12 +57,17 @@ enum VersionedAlephJustification {
 }
 
 fn encode_with_version(version: Version, mut payload: Vec<u8>) -> Vec<u8> {
-    let mut result = version.encode();
+    let mut result =
+        Vec::with_capacity(size_of::<Version>() + size_of::<ByteCount>() + payload.len());
+    version.encode_to(&mut result);
     // This will produce rubbish if we ever try encodings that have more than u16::MAX bytes. We
     // expect this won't happen, since we will switch to proper multisignatures before proofs get
     // that big.
-    let num_bytes = payload.len() as ByteCount;
-    result.append(&mut num_bytes.encode());
+    payload
+        .len()
+        .try_into()
+        .unwrap_or(u16::MAX)
```

Comment:
It's a bit confusing to return data that is potentialy corrupt if it's larger than `MAX`.

Reply:
I do not see any better way to do this. We do not want to panic here and we also cannot return an error here. This way it will work for everyone who knows the version, and people who do not know the version will also be safe.

Besides that, situation when this exceeds `u16::MAX` will likely never happen.

We have this maximum, so that no one spams a node not knowing the version with creation of huge buffers.

Reply:
What about checking this condition first and returning empty `Vec` then?

Reply:
This could potentially break our code if we incorrectly estimate max size of justification. We do not want to return empty vector from this function in case if something is a correct justification

Reply:
Current version changes format if payload::size() > u16::MAX. Anyway, assumption here is that it shouldn't happen. But if, returned vec might be corrupted in that case (truncated). Leave as it is now, or don't pretend that returned data has any meaning (corrupted) and return empty or error. I would prefer early exit and returning error or actually not allowing to build payload that is potentially too big in the first place (but it is too late for that design). Decision is yours. You can resolve this comment.

Reply:
Actually the `extend_from_slice` will expand the underlying vector if it's too small. Meaning, we would create a message that is bigger than the message limit.

Reply:
Yep, and we want to create a message that is too big in this case. So that it works for everyone who has this version. In decode we do not check size if we know the version.